[#26, #27, #66] 매장 등록, 구이름 조회, 거리 순 조회 API 개발

tinajeong opened the pull request:
### 🍀 목적
- 매장 도메인 구현

### 🌹 추가사항(있다면 적고 없다면 적지 않는다.)
- 매장 등록
- 매장 아이디로 조회
- 매장 거리 순 조회

tinajeong committed:
feat : 매장 기능 구현
- 매장 추가, 아이디로 조회, 구이름으로 조회, 현재 위치 기준으로 조회 기능 구현
- repository, 통합테스트 작성

tinajeong committed:
feat : 매장 관련 코드 리팩토링
- StoreResponse Inner class 의미 중복 제거
- DefaultStoreService 메소드 별로 @Transactional 추가
- StoreRepository findAllByDistance 오타 수정
- StoreCreateRequest의 double 타입 @NotNull로 검증

devYSK commented on src/main/java/com/prgrms/bdbks/domain/store/api/StoreController.java:
```diff
@@ -0,0 +1,59 @@
+	}
+
+	@GetMapping(produces = MediaType.APPLICATION_JSON_VALUE)
+	public ResponseEntity<List<StoreResponse.Information>> findStoresByDistrict(
```
List 를 한번 Wrapping해주는 dto를 하나 만들면 어떨까요?

tinajeong replied:
반영하겠습니다~

devYSK commented on src/main/java/com/prgrms/bdbks/domain/store/api/StoreController.java:
```diff
@@ -0,0 +1,59 @@
+
+	@GetMapping(value = "/location", produces = MediaType.APPLICATION_JSON_VALUE)
+	public ResponseEntity<List<StoreResponse.Information>> findStoresByLocation(
+		@RequestParam Double latitude, @RequestParam Double longitude) {
```
parameter에 대한 validation도 필요할 것 같습니다!

tinajeong replied:
추가했습니다!

devYSK commented on src/main/java/com/prgrms/bdbks/domain/store/entity/Store.java:
```diff
@@ -75,4 +87,24 @@ private void validationName(String name) {
+
+	@Component
+	protected static class PointConverter implements AttributeConverter<Point, String> {
+		static WKTReader wktReader = new WKTReader();
```
자주 쓰이는거같은데, 한곳에서 선언해서 놓고 쓰면 좋을것같아요~

Bean이니까 JpaConfig같은 외부에 선언하는건 어떨까요?

devYSK commented on src/main/java/com/prgrms/bdbks/domain/store/repository/StoreRepository.java:
```diff
@@ -0,0 +1,31 @@
+import com.prgrms.bdbks.global.util.Location;
+
+@Repository
+public interface StoreRepository extends JpaRepository<Store, String> {
```
LGTM

nativequery 고수십니다.

devYSK commented on src/main/java/com/prgrms/bdbks/domain/store/service/DefaultStoreService.java:
```diff
@@ -0,0 +1,95 @@
+	private final StoreRepository storeRepository;
+	private final StoreMapper storeMapper;
+
+	@Override
```
readonly 트랜잭션 선언이 빠진것같아요~

tinajeong replied:
수정하겠습니다!

devYSK commented on src/main/java/com/prgrms/bdbks/global/util/Location.java:
```diff
@@ -0,0 +1,35 @@
+		this.longitude = longitude;
+
+		try {
+			this.point = (Point)new WKTReader().read(
```
공백 위치가 뭔가 애매한거같아요!

devYSK commented on src/test/java/com/prgrms/bdbks/domain/store/repository/StoreRepositoryTest.java:
```diff
@@ -0,0 +1,79 @@
+		assertEquals(2L, storeRepository.count());
+	}
+
+	@Test
+	@DisplayName("조회 - 좌표간 거리 환산에 성공 한다.")
+	void distance_test() {
+
+		Location location = new Location(37.493657, 127.013772);
+		List<Integer> distances = storeRepository.findDistance(location);
+
+		assertTrue(distances.stream().allMatch(x -> x != null));
+	}
+
+	@Test
+	@DisplayName("조회 - 거리가 가까운 순으로 조회하는데 성공 한다.")
+	void short_distance_test() {
```
정확하진않지만 근삿값 숫자로 범위 비교라도 해서 테스트 코드에서 보이게 하는건 어떨까요 !

tinajeong replied:
반영하겠습니다~

devYSK reviewed (approved):
어려운 기능 구현하시느냐 고생 많으셨습니다!

tinajeong committed:
feat : findByUserId,findByLoginId 인터페이스 추가

tinajeong committed:
refact : 입출력 값 관련 리팩토링
- pointInformation -> point
- `@Transactional(readOnly = true)` 추가
- 위경도 입력 Validation 추가

tinajeong committed:
Merge branch 'develop' into feature/store

tinajeong committed:
chore : resolve conflict
